src/devices/watts_thermostat.c

```diff
@@ -0,0 +1,199 @@
+    Level estimates [high, low]:  15961,    150
+    RSSI: -0.1 dB SNR: 20.3 dB Noise: -20.4 dB
+    Frequency offsets [F1, F2]:   17869,      0     (+545.3 kHz, +0.0 kHz)
+    Guessing modulation: Pulse Width Modulation with sync/delimiter
+    view at https://triq.org/pdv/#AAB104FFFF025A0105FFFF8291A291A1A291A29291A29291A291A1A1A2929291A29291A1A1A1A1A29291A291A1A1A1A291A1A29291A1A29291A29291A1A2929291A355
+    Attempting demodulation... short_width: 288, long_width: 630, reset_limit: 577, sync_width: 291096
+    Use a flex decoder with -X 'n=name,m=OOK_PWM,s=288,l=630,r=577,g=0,t=0,y=291096'
+    pulse_slicer_pwm(): Analyzer Device
+    bitbuffer:: Number of rows: 1
+    [00] {54} 5a 4b 89 f2 f6 64 c4
+
+
+Data Layout:
+
+    10100101   1011010001110110   1000   100100001   000011000   10101011
+    preamble   id                 flags  temp         setpoint   chksum
+
+    All fields need reflection, possibly easier to just reverse the whole
+    row first. The only flag found is PAIRING (0b0001). Chksum seems to be
+    additive.
+
+Raw data:
+
+    {54}5ab24971f79994
+    {54}5ab24971f79994
+    {54}5ab249f1f79b94
+    {54}5ab249f1f79b94
+    {54}5ab249f9f79854
+    {54}5ab249f5f79a54
+    {54}5ab249f68f998c
+    {54}5ab249f98f9a4c
+    {54}5ab249f58b9a4c
+    {54}5ab249fb8f9acc
+
+    https://tinyurl.com/2z5jtfuu
+
+    Format string:
+    ID:^16d FLAGS:^4b TEMP:^9d SETP:^9d CHK:^8d
+
+Decoded example:
+
+    ID:28205 FLAGS:0001 TEMP:265 SETP:048 CHK:214
+
+*/
+
+static int watts_thermostat_decode(r_device *decoder, bitbuffer_t *bitbuffer)
+{
+    int ret                          = 0;
+
+    if (bitbuffer->num_rows != 1) {
+        decoder_logf(decoder, 2, __func__, "Only expect a single row. num_rows=%d", bitbuffer->num_rows);
+        ret = DECODE_ABORT_EARLY;
+        return ret;
+    }
+
+    uint8_t const preamble_pattern = 0xa5;
+
+    for (unsigned row = 0; row < bitbuffer->num_rows; ++row) {
+        // we expect 54 bits
+        if (bitbuffer->bits_per_row[row] != 54) {
+            decoder_log(decoder, 2, __func__, "Length check fail");
+            ret = DECODE_ABORT_LENGTH;
+            continue;
+        }
+
+        bitbuffer_invert(bitbuffer);
+
+        unsigned pos = bitbuffer_search(bitbuffer, row, 0, &preamble_pattern, 8);
+
+        if (pos >= bitbuffer->bits_per_row[row]) {
+            decoder_log(decoder, 2, __func__, "Preamble not found");
+            ret = DECODE_ABORT_EARLY;
+            continue;
+        }
+
+        pos += 8;
+
+        uint8_t id_raw[2] = {0};
+        bitbuffer_extract_bytes(bitbuffer, row, pos, id_raw, 16);
+        unsigned id = (reverse8((id_raw[1])) << 8) | reverse8((id_raw[0]));
+
+        pos += 16;
+
+        uint8_t PAIRING = 1;
+        uint8_t flags   = {0};
+        bitbuffer_extract_bytes(bitbuffer, row, pos, &flags, 4);
+        flags = reverse8(flags);
+        unsigned pairing = flags & PAIRING;
+
+        pos += 4;
+
+        uint8_t temp_raw[2] = {0};
```

Comment:
I realize it seems clean to init, but presumably we are extracting into it.

Reply:
You are probably right then.   But, I find extracting 4 bits to be odd.   I usually see pulling the bytes out to variables and then extracting with bit ops.   I personally like to see if all other bits are the expected values and add some diagnostic-type data elements when not, to help others figure out more.